### ISSUE TITLE
Visual Studio 2013 and Windows Phone 8.1

### ISSUE THREAD

**Issue by gameboy9** [ABB-Austin/cordova-plugin-indexeddb-async#1]
I'm trying to get this plugin to work with Windows Phone 8.1, but I'm royally struggling because I can't seem to figure out how to specify the extra flag required("cordova build windows --archs="x86 arm" -- --phone") to get the extensions to work in Visual Studio 2013.  Attempting to build in the default "debug/Windows Phone (Universal)" configuration results in this error:

"The processor architecture of the project being built "Any CPU" is not supported by the referenced SDK "Microsoft.VCLibs, Version=12.0". Please consider changing the targeted processor architecture of your project (in Visual Studio this can be done through the Configuration Manager) to one of the architectures supported by the SDK: "x86, ARM"." (there is a second bug for SQLite.WP81 as well)

Do you know how I can fix that in Visual Studio 2013 so I can properly debug my programs?  Thank you for any help you have in advance.

**Comment by JamesMessinger** [comment]
Yeah, having to specify the processor architecture ( `--archs="x86 arm"`) is a bit of a pain.  The reason this is necessary is because [the async WebSQL plugin](https://github.com/Thinkwise/cordova-plugin-websql) is specifically compiled for ARM and x86, rather than as a universal binary.   This means that you _can't_ use Visual Studio's default "Any CPU" configuration, since it tries to build a universal binary.

To _try_ to make this process less painful, we run [this script](https://github.com/Thinkwise/cordova-plugin-websql/blob/master/scripts/after_plugin_install.js) when you first install the plugin via `cordova plugin add`.   The script modifies your Visual Studio Solution (`.sln` file) and Windows Phone Project (`.jsproj` file), so that the "Any CPU" and "x64" platforms actually build x86 binaries rather than universal/x64 binaries.   If all goes well, you should be good.  But if it doesn't, then you may have to manually edit your build configurations in Visual Studio.

To do that, open Visual Studio and go to "Build -> Configuration Manager" in the menu.  Make sure your "Any CPU" and "x64" platforms are set to "x86" for the CordovaApp.Phone project.  Be sure to check this for "Debug" and "Release" configurations.

![screen shot 2015-07-16 at 10 37 33 am](https://cloud.githubusercontent.com/assets/3453903/8727710/c8e4cbd2-2ba6-11e5-9180-301159ca00db.png)

Another option is to just remove the "Any CPU" and "x64" platforms:

![screen shot 2015-07-16 at 10 38 05 am](https://cloud.githubusercontent.com/assets/3453903/8727758/10c9f76a-2ba7-11e5-8bdc-c6dd768ca02a.png)

**Comment by gameboy9** [comment]
Thanks for your reply.  It turns out that Any CPU, ARM, x64, or x86 are solution platforms for my project, seen below:  

![image](https://cloud.githubusercontent.com/assets/13355791/8728119/6e03b556-2bb1-11e5-8f53-891cd1ee7499.png)

I, theoretically, could change the platform to Windows-ARM or Windows-x86, but if I do that, I can only debug via a "remote machine", not through a phone emulator or definitely my physical Windows Phone.  I'm using a "Tools for Apache Cordova" project.  Should I be using a different type of project altogether?  (Although I was hoping that this project would work for Android and iOS as well...)

Thanks again.

**Comment by JamesMessinger** [comment]
Hmmm... I've never used the "Tools for Apache Cordova", so I don't know what differences it has from a normal Cordova project (that is, one created using [`cordova create`](https://cordova.apache.org/docs/en/edge/guide_cli_index.md.html#The%20Command-Line%20Interface_create_the_app)).   I'm sure there's a way to make it all play nice together though.  There's got to be a way to configure the processor platform for Windows Phone separately from iOS, Android, etc.

**Comment by gameboy9** [comment]
OK, so I created a batch file so all of the plugins I need, etc. are copied and should be ready to go:

> @echo off
> ECHO Y | rmdir x /s
> call cordova create x x.WP8 xWP8
> cd x
> call cordova plugin add cordova-plugin-inappbrowser
> call cordova plugin add cordova-plugin-indexeddb-async
> call cordova plugin add cordova-plugin-websql-async
> ...
> rem copying files from VS project to WP8 project.
> copy "..." "...\x\www"
> copy "...\scripts" "...\x\www\scripts"
> copy "...\css" "...\x\www\css"
> copy "...\images" "...\x\www\images"
> call cordova platform add wp8
> call cordova build wp8 --debug --archs="x86 arm" -- --phone
> call cordova run wp8

I'm afraid that it still doesn't work.  Is there something I could be missing?  Is there a way I can print out console messages to my windows phone, since I can't get messages through the CLI (that I know of)?  Thanks again!

**Comment by JamesMessinger** [comment]
So the `call cordova build` command succeeds, and then you get runtime errors when running the app on the phone?   Does the app even launch?  Do you get an error message?

**Comment by gameboy9** [comment]
OK, I finally have it where I can finally give you some more information.  I have a series of console.logs running, and I get it to where it's running "upgrade needed", but then it runs "onerror", with no information.  However, with a little bit of "console.log" tracing, I've discovered that WP8 doesn't like this line for some reason:

...
        console.log("running upgrade needed2");
        if (event.oldVersion < 1) {
            console.log("running upgrade needed2.1");
...
}
        console.log("running upgrade needed4");
...

"needed2" prints, "needed2.1", right after the if statement, and "needed4", right after the entire if block, does not.

This totally works fine if I try this in Android.  Is this a WP8 issue by chance?  Thanks again!

**Comment by gameboy9** [comment]
There's one other piece of information I should share:  Despite me adding the appropriate plugins: 
cordova plugin add cordova-plugin-indexeddb-async
cordova plugin add cordova-plugin-websql-async

When I try to build using "call cordova build wp8 --debug --archs="x86 arm" -- --phone", those two plugins don't wind up in the plugins folder.  I wonder why that is...

**Comment by gameboy9** [comment]
OK, never mind this, that was a programming error on my side.  Let me do some more digging... thanks!

**Comment by gameboy9** [comment]
OK, so here's what I did:  I removed all of the plugins in my Cordova project, then readded them (with numerous updates).  After this, I was able to build the project via the "debug/Windows Phone 8" configuration.  This deploys to my device correctly, but I don't get any debug capabilities(i.e. console.log).  However, since a csproj and sln file is created on the build, I can load that and get debug capability that way.  Anyway, after some testing, it looks like IndexedDB now works with your plugins correctly, with no corrections in code required on your side.  Hallelujah... it took me over a week to figure this out!  :(

Thanks for all of your help.
